### PR TITLE
Add search engine support for files

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Add search engine support for files

#### Related PRs
Enterprise: mattermost/enterprise#815

### CODE REVIEW

File: store/sqlstore/file_info_store.go [hunk 1]
```diff
@@ -51,6 +55,9 @@ func (fs SqlFileInfoStore) createIndexesIfNotExists() {
 	fs.CreateIndexIfNotExists("idx_fileinfo_create_at", "FileInfo", "CreateAt")
 	fs.CreateIndexIfNotExists("idx_fileinfo_delete_at", "FileInfo", "DeleteAt")
 	fs.CreateIndexIfNotExists("idx_fileinfo_postid_at", "FileInfo", "PostId")
+	fs.CreateIndexIfNotExists("idx_fileinfo_extension_at", "FileInfo", "Extension")
+	fs.CreateFullTextIndexIfNotExists("idx_fileinfo_name_txt", "FileInfo", "Name")
+	fs.CreateFullTextIndexIfNotExists("idx_fileinfo_content_txt", "FileInfo", "Content")
```
Comment (ethervoid) [inline store/sqlstore/file_info_store.go:60]: One question. That includes the content of the file? I mean is the whole content of a file?

Comment (jespino) [reply store/sqlstore/file_info_store.go:60]: Not really, the Content is a new field added that (if the setting is enabled) is going to cointain the extracted text from the file. For example, if this is a pdf that have the text "hello world" in it, the `Content` field is going to contain the "hello world" text, not the binary data.

Comment (ethervoid) [reply store/sqlstore/file_info_store.go:60]: What would be inside if the file has 8mb of size? All the extracted document?

Comment (jespino) [reply store/sqlstore/file_info_store.go:60]: All the extracted plain-text (but only for supported files, and always behind a configuration flag).

File: store/sqlstore/file_info_store.go [hunk 2]
```diff
@@ -393,3 +420,176 @@ func (fs SqlFileInfoStore) PermanentDeleteByUser(userId string) (int64, error) {
...
+			onDateStart, onDateEnd := params.GetOnDateMillis()
+			query = query.Where(sq.Expr("FI.CreateAt BETWEEN ? AND ?", strconv.FormatInt(onDateStart, 10), strconv.FormatInt(onDateEnd, 10)))
+		} else {
+			if len(params.ExcludedDate) > 0 {
+				excludedDateStart, excludedDateEnd := params.GetExcludedDateMillis()
+				query = query.Where(sq.Expr("FI.CreateAt NOT BETWEEN ? AND ?", strconv.FormatInt(excludedDateStart, 10), strconv.FormatInt(excludedDateEnd, 10)))
+			}
+
+			if len(params.AfterDate) > 0 {
+				afterDate := params.GetAfterDateMillis()
```
Comment (ethervoid) [inline store/sqlstore/file_info_store.go:489]: As a side comment, I've found that we're [doing fallback to date.Now() + 1 day](https://github.com/mattermost/mattermost-server/blob/master/model/search_params.go#L41) if the `params.AfterDate` fails to be parsed. That could be weird

**Review by ethervoid — COMMENTED** [review]
Initial review batch

File: model/file_info_list.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+import (
+	"encoding/json"
+	"io"
+	"sort"
+)
+
+type FileInfoList struct {
+	Order          []string             `json:"order"`
+	FileInfos      map[string]*FileInfo `json:"file_infos"`
```
[the remark this answers is not in the document]
Comment (jespino) [reply model/file_info_list.go:14]: That is out of the scope of this PR, is already complex enough, I think we can address that as part of other ticket, but now is replicating the behavior in other parts of the code.

File: model/file_info_list.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+	if o.FileInfos == nil {
+		o.FileInfos = make(map[string]*FileInfo)
+	}
+}
+
+func (o *FileInfoList) AddOrder(id string) {
+
+	if o.Order == nil {
+		o.Order = make([]string, 0, 128)
```
[the remark this answers is not in the document]
Comment (jespino) [reply model/file_info_list.go:58]: Same problem here, this is consistent with other part of the code, I think we can create a ticket to remove this magic numbers in any search list result that we have.

File: model/file_info_list.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+func (o *FileInfoList) AddFileInfo(fileInfo *FileInfo) {
+
+	if o.FileInfos == nil {
+		o.FileInfos = make(map[string]*FileInfo)
+	}
+
+	o.FileInfos[fileInfo.Id] = fileInfo
+}
+
+func (o *FileInfoList) UniqueOrder() {
```
[the remark this answers is not in the document]
Comment (jespino) [reply model/file_info_list.go:73]: Again, consistency with other parts of the code. Maybe is not needed, but is heavily based on the `model/post_list.go` and I prefer to keep them consistent and change this in other tickets.

File: model/file_info_list.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	})
+}
+
+func (o *FileInfoList) Etag() string {
+
+	id := "0"
+	var t int64 = 0
+
+	for _, v := range o.FileInfos {
+		if v.UpdateAt > t {
```
[the remark this answers is not in the document]
Comment (jespino) [reply model/file_info_list.go:110]: Consistency again ;)

File: model/file_info_list.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+	if len(o.Order) > 0 {
+		orderId = o.Order[0]
+	}
+
+	return Etag(orderId, id, t)
+}
+
+func FileInfoListFromJson(data io.Reader) *FileInfoList {
+	var o *FileInfoList
+	json.NewDecoder(data).Decode(&o)
```
[the remark this answers is not in the document]
Comment (jespino) [reply model/file_info_list.go:129]: Consistency, we are not checking the error in any other model.

File: model/file_info_search_results.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	b, err := json.Marshal(o)
+	if err != nil {
+		return ""
+	}
+	return string(b)
+}
+
+func FileInfoSearchResultsFromJson(data io.Reader) *FileInfoSearchResults {
+	var o *FileInfoSearchResults
+	json.NewDecoder(data).Decode(&o)
```
[the remark this answers is not in the document]
Comment (jespino) [reply model/file_info_search_results.go:35]: Same problem.

File: services/searchengine/bleveengine/indexer/indexing_job.go [hunk 9]
```diff
@@ -355,6 +355,30 @@ func (worker *BleveIndexerWorker) BulkIndexPosts(posts []*model.PostForIndexing,
...
+func (worker *BleveIndexerWorker) BulkIndexFiles(files []*model.FileForIndexing, progress IndexingProgress) (int64, *model.AppError) {
+	lastCreateAt := int64(0)
+	batch := worker.engine.FileIndex.NewBatch()
+
+	for _, file := range files {
+		if file.DeleteAt == 0 {
+			searchFile := bleveengine.BLVFileFromFileForIndexing(file)
+			batch.Index(searchFile.Id, searchFile)
+		} else {
+			batch.Delete(file.Id)
```
[the remark this answers is not in the document]
Comment (jespino) [reply services/searchengine/bleveengine/indexer/indexing_job.go:367]: Yes, still, this is exactly how was behaving other implementations of this same thing :)

File: store/searchlayer/file_info_layer.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+					return
+				}
+				post, postErr := s.rootStore.Post().GetSingle(file.PostId)
+				if postErr != nil {
+					mlog.Error("Couldn't get post for file for SearchEngine indexing.", mlog.String("post_id", file.PostId), mlog.String("search_engine", engineCopy.GetName()), mlog.String("file_info_id", file.Id), mlog.Err(postErr))
+					return
+				}
+
+				if err := engineCopy.IndexFile(file, post.ChannelId); err != nil {
+					mlog.Error("Encountered error indexing file", mlog.String("file_info_id", file.Id), mlog.String("search_engine", engineCopy.GetName()), mlog.Err(err))
```
[the remark this answers is not in the document]
Comment (jespino) [reply store/searchlayer/file_info_layer.go:32]: Again, it is this way in the other implementations.

File: store/searchlayer/file_info_layer.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+		}
+	}
+}
+
+func (s SearchFileInfoStore) deleteFileIndex(fileID string) {
+	for _, engine := range s.rootStore.searchEngine.GetActiveEngines() {
+		if engine.IsIndexingEnabled() {
+			runIndexFn(engine, func(engineCopy searchengine.SearchEngineInterface) {
+				if err := engineCopy.DeleteFile(fileID); err != nil {
+					mlog.Error("Encountered error deleting file", mlog.String("file_info_id", fileID), mlog.String("search_engine", engineCopy.GetName()), mlog.Err(err))
```
[the remark this answers is not in the document]
Comment (jespino) [reply store/searchlayer/file_info_layer.go:45]: same answer ;)

File: store/searchlayer/file_info_layer.go [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+		}
+	}
+}
+
+func (s SearchFileInfoStore) deleteFileIndexForUser(userID string) {
+	for _, engine := range s.rootStore.searchEngine.GetActiveEngines() {
+		if engine.IsIndexingEnabled() {
+			runIndexFn(engine, func(engineCopy searchengine.SearchEngineInterface) {
+				if err := engineCopy.DeleteUserFiles(userID); err != nil {
+					mlog.Error("Encountered error deleting files for user", mlog.String("user_id", userID), mlog.String("search_engine", engineCopy.GetName()), mlog.Err(err))
```
[the remark this answers is not in the document]
Comment (jespino) [reply store/searchlayer/file_info_layer.go:58]: same

File: store/sqlstore/file_info_store.go [hunk 13]
```diff
@@ -393,3 +420,176 @@ func (fs SqlFileInfoStore) PermanentDeleteByUser(userId string) (int64, error) {
 
 	return rowsAffected, nil
 }
+
+func (fs SqlFileInfoStore) Search(paramsList []*model.SearchParams, userId, teamId string, page, perPage int) (*model.FileInfoList, error) {
+	// Since we don't support paging for DB search, we just return nothing for later pages
+	if page > 0 {
+		return model.NewFileInfoList(), nil
```
[the remark this answers is not in the document]
Comment (jespino) [reply store/sqlstore/file_info_store.go:427]: Again, consistent with other implementation in the codebase.

File: store/sqlstore/file_info_store.go (hunk 2 above)
Comment (jespino) [reply store/sqlstore/file_info_store.go:489]: Consistent with the other code :)

**Review by mgdelacroix — COMMENTED** [review]
This is going to be awesome! Some comments noted

**Review by mgdelacroix — APPROVED** [review]
LGTM! Thanks for applying the changes! :+1:

**Review by ethervoid — APPROVED** [review]
LGTM! Really amazing stuff. I'm eager to see it working :D

**Review by lindalumitchell — APPROVED** [review]

**Pull request merged**